File: examples/LMemLoopback/client/py/AdvancedStatic/LMemLoopbackClient.py
#!/usr/bin/env python

import sys, glob
import random
sys.path.append('../gen-py')

from com.maxeler.LMemLoopback import LMemLoopbackService
from com.maxeler.LMemLoopback.ttypes import*

from thrift import Thrift
from thrift.transport import TSocket
from thrift.transport import TTransport
from thrift.protocol import TBinaryProtocol

def check(size, out_data, in_a, in_b):
    """Check if out_data = in_a + in_b"""
    status = 0
    for i in range(0, size):
        if out_data[i] != in_a[i] + in_b[i]:
            print str(out_data[i]) + " != " + str(in_a[i]) + " + " + str(in_b[i])
            status = 1
    return status

def LMemLoopback_dfe(size, in_a, in_b):
    """LMemLoopback DFE implementation."""
    try:

        # Make socket
        transport = TSocket.TSocket('localhost', 9090)

        # _buffering is critical. Raw sockets are very slow
        transport = TTransport.TBufferedTransport(transport)

        # Wrap in a protocol
        protocol = TBinaryProtocol.TBinaryProtocol(transport)

        # Create a client to use the protocol encoder
        client = LMemLoopbackService.Client(protocol)

        # Connect!
        transport.open()

        size_bytes = size * 4

        # Initialize maxfile
        max_file = client.LMemLoopback_init()

        # Load DFE
        max_engine = client.max_load (max_file, '*')    

        # Allocate and send input streams to server
        address_in_a = client.malloc_int32_t(size)
        client.send_data_int32_t(address_in_a, in_a)

        address_in_b = client.malloc_int32_t(size)
        client.send_data_int32_t(address_in_b, in_b)

        # Allocate memory for output stream on server
        address_out_data = client.malloc_int32_t(size)

        print "Loading DFE memory."
        actions_writeLMem = LMemLoopback_writeLMem_actions_t_struct(0, size_bytes, address_in_a)
        address_actions_writeLMem = client.send_LMemLoopback_writeLMem_actions_t(actions_writeLMem)
        client.LMemLoopback_writeLMem_run(max_engine, address_actions_writeLMem)

        actions_writeLMem = LMemLoopback_writeLMem_actions_t_struct(size_bytes, size_bytes, address_in_b)
        address_actions_writeLMem = client.send_LMemLoopback_writeLMem_actions_t(actions_writeLMem)
        client.LMemLoopback_writeLMem_run(max_engine, address_actions_writeLMem)

        print "Running DFE."
        actions = LMemLoopback_actions_t_struct(size)
        address_actions = client.send_LMemLoopback_actions_t(actions)
        client.LMemLoopback_run(max_engine, address_actions)

        print "Reading DFE memory."
        actions_readLMem = LMemLoopback_readLMem_actions_t_struct(2 * size_bytes, size_bytes, address_out_data)
        address_actions_readLMem = client.send_LMemLoopback_readLMem_actions_t(actions_readLMem)
        client.LMemLoopback_readLMem_run(max_engine, address_actions_readLMem)

        # Unload DFE
        client.max_unload(max_engine)

        # Get output stream from server
        out_data = client.receive_data_int32_t(address_out_data, size)

        # Free allocated memory for streams on server
        client.free(address_in_a)
        client.free(address_in_b)
        client.free(address_out_data)

        # Free allocated maxfile data
        client.LMemLoopback_free()

        # Close!
        transport.close()

    except Thrift.TException, thrift_exceptiion:
        print '%s' % (thrift_exceptiion.message)

    return out_data

def test():
    '''
    Calls LMemLoopbackDFE and LMemLoopbackCPU
    and checks if they return the same result.
    '''
    # Input
    size = 384

    in_a = [i for i in range(size)]
    in_b = [size - i for i in range(size)]

    # DFE Output
    out_data = LMemLoopback_dfe(size, in_a, in_b)

    # Checking results
    if check(size, out_data, in_a, in_b):
        print "Test failed."
    else:
        print "Test passed!"

if __name__ == '__main__':
    test()

File: examples/LMemLoopback/client/hs/AdvancedStatic/LMemLoopbackClient.hs
module Main where

import qualified LMemLoopbackService_Client as Client
import LMemLoopback_Types

import Thrift
import Thrift.Protocol.Binary
import Thrift.Server
import Thrift.Transport
import Thrift.Transport.Handle

import Control.Exception
import Data.Either
import Data.Int
import Data.List
import Data.Maybe
import Data.Time
import Data.Text.Lazy
import Data.Vector
import Network
import System.Exit
import System.Random
import Text.Printf

getRight           :: Either left right -> right
getRight (Right x) = x

lMemLoopbackCPU :: [Int32] -> [Int32] -> [Int32]
lMemLoopbackCPU [] []         = []
lMemLoopbackCPU (a:inA) (b:inB) = (a + b) : (lMemLoopbackCPU (inA) (inB) )

check :: [Int32] -> [Int32] -> Int -> Int -> [Int]
check [] [] start end = []
check (x:outDFE) (y:outCPU) start end
	| (x == y)		=  check outDFE outCPU (start+1) end
	| otherwise 	= (start) : check outDFE outCPU (start+1) end

printErrors :: [Int] -> [Int32] -> [Int32] -> String -> String
printErrors [] xs ys output                 = output
printErrors (i:is) xs ys output     = printErrors is xs ys (output Data.List.++ "Output data @ " Data.List.++ (show i) Data.List.++ " = " Data.List.++ (show (xs!!i))  Data.List.++ " (expected " Data.List.++ (show (ys!!i))  Data.List.++ ")\n")

main = do
    startTime <- getCurrentTime
    startDFETime <- getCurrentTime
    
    -- Make socket
    transport <- hOpen ("localhost", PortNumber 9090)

    -- Wrap in a protocol
    let protocol = BinaryProtocol transport

    -- Create a client to use the protocol encoder
    let client = (protocol, protocol)
    stopTime <- getCurrentTime
    putStrLn ("Creating a client and opening connection:\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Generate input data
    startTime <- getCurrentTime
    let size = 384
    let sizeBytes = size * 4
    let inA = [0 .. (size-1)]
    let inB = [size - a | a <- inA]
    stopTime <- getCurrentTime
    putStrLn ("Generating input data:\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Initialize maxfile
    startTime <- getCurrentTime
    e <- try (Client.lMemLoopback_init client) :: IO (Either SomeException Int64)
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    let maxfile = getRight e
    stopTime <- getCurrentTime
    putStrLn ("Initializing maxfile:\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Load DFE
    startTime <- getCurrentTime
    e <- try (Client.max_load client maxfile (pack "*")) :: IO (Either SomeException Int64)
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    let engine = getRight e
    stopTime <- getCurrentTime
    putStrLn ("Loading DFE:\t\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Allocate and send input streams to server
    startTime <- getCurrentTime
    e <- try (Client.malloc_int32_t client (fromIntegral size)) :: IO (Either SomeException Int64)
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    let address_inA = getRight e

    e <- try (Client.send_data_int32_t client address_inA (fromList inA)) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    e <- try (Client.malloc_int32_t client (fromIntegral size)) :: IO (Either SomeException Int64)
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    let address_inB = getRight e

    e <- try (Client.send_data_int32_t client address_inB (fromList inB)) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
        
    stopTime <- getCurrentTime
    putStrLn ("Sending input data:\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Allocate memory for output stream on server
    startTime <- getCurrentTime

    e <- try (Client.malloc_int32_t client (fromIntegral size)) :: IO (Either SomeException Int64)
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    let address_dataOut = getRight e

    stopTime <- getCurrentTime
    putStrLn ("Allocating memory for output stream on server:\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Writing to LMem 
    startTime <- getCurrentTime
    let actions_lmemA = LMemLoopback_writeLMem_actions_t_struct (Just (fromIntegral 0)) (Just (fromIntegral sizeBytes)) (Just (fromIntegral address_inA))
    e <- try (Client.send_LMemLoopback_writeLMem_actions_t client actions_lmemA) :: IO (Either SomeException Int64)
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    let address_actions_lmemA = getRight e

    e <- try (Client.lMemLoopback_writeLMem_run client engine address_actions_lmemA) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    let actions_lmemB = LMemLoopback_writeLMem_actions_t_struct (Just (fromIntegral sizeBytes)) (Just (fromIntegral sizeBytes)) (Just (fromIntegral address_inB))
    e <- try (Client.send_LMemLoopback_writeLMem_actions_t client actions_lmemB) :: IO (Either SomeException Int64)
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    let address_actions_lmemB = getRight e

    e <- try (Client.lMemLoopback_writeLMem_run client engine address_actions_lmemB) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    stopTime <- getCurrentTime
    putStrLn ("Writing to LMem:\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Action default
    startTime <- getCurrentTime
    let action = LMemLoopback_actions_t_struct (Just (fromIntegral size))
    e <- try (Client.send_LMemLoopback_actions_t client action) :: IO (Either SomeException Int64)
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    let address_action = getRight e

    e <- try (Client.lMemLoopback_run client engine address_action) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    stopTime <- getCurrentTime
    putStrLn ("LMemLoopback time:\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Reading from LMem
    startTime <- getCurrentTime
    let actions_lmemRead = LMemLoopback_readLMem_actions_t_struct (Just (fromIntegral (2 * sizeBytes))) (Just (fromIntegral sizeBytes)) (Just (fromIntegral address_dataOut))
    e <- try (Client.send_LMemLoopback_readLMem_actions_t client actions_lmemRead) :: IO (Either SomeException Int64)
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    let address_actions_lmemRead = getRight e

    e <- try (Client.lMemLoopback_readLMem_run client engine address_actions_lmemRead) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    stopTime <- getCurrentTime
    putStrLn ("Reading from LMem:\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Unload DFE
    startTime <- getCurrentTime
    e <- try (Client.max_unload client engine) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    stopTime <- getCurrentTime
    putStrLn ("Unloading DFE:\t\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Get output stream from server
    startTime <- getCurrentTime
    dataOutDFE <- Client.receive_data_int32_t client address_dataOut (fromIntegral size)
    stopTime <- getCurrentTime
    putStrLn ("Getting output stream:\t(size = " Data.List.++ (show (size * 32)) Data.List.++ " bit)\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Free allocated memory for streams on server
    startTime <- getCurrentTime
    e <- try (Client.free client address_inA) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    e <- try (Client.free client address_inB) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    e <- try (Client.free client address_dataOut) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()
    
    stopTime <- getCurrentTime
    putStrLn ("Freeing allocated memory for streams on server:\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Free allocated maxfile data
    startTime <- getCurrentTime
    e <- try (Client.lMemLoopback_free client) :: IO (Either SomeException ())
    case e of
        Left ex -> putStrLn $ "Caught exception: " Data.List.++ show ex
        Right ex -> return ()

    -- Close!
    startTime <- getCurrentTime
    tClose transport
    stopTime <- getCurrentTime
    putStrLn ("Closing connection:\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    stopTime <- getCurrentTime
    putStrLn ("DFE LMemLoopback total time:\t\t\t" Data.List.++ (show (diffUTCTime stopTime startDFETime)))

    -- CPU Output
    startTime <- getCurrentTime
    let dataOutCPU = lMemLoopbackCPU inA inB
    stopTime <- getCurrentTime
    putStrLn ("CPU LMemLoopback total time::\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    -- Checking results
    startTime <- getCurrentTime
    let errors = check (toList dataOutDFE) dataOutCPU 0 (fromIntegral size)
    stopTime <- getCurrentTime
    putStrLn ("Checking results:\t\t\t\t" Data.List.++ (show (diffUTCTime stopTime startTime)))

    if ((Data.List.length errors)== 0)
	then putStrLn ("Test successful!")
        else do putStr (printErrors errors (toList dataOutDFE) dataOutCPU [])
                putStrLn ("Test failed "  Data.List.++ show (Data.List.length errors)  Data.List.++ " times!")
                exitWith $ ExitFailure (-1)


